src/auth/Crypto.cc

```diff
@@ -231,22 +212,30 @@ class CryptoAESKeyHandler : public CryptoKeyHandler {
     ivItem.data = (unsigned char*)CEPH_AES_IV;
     ivItem.len = sizeof(CEPH_AES_IV);
 
+    SECItem *param = nullptr;
     param = PK11_ParamFromIV(mechanism, &ivItem);
     if (!param) {
       err << "cannot set NSS IV param: " << PR_GetError();
       return -1;
     }
 
+    enc_ctx.reset(PK11_CreateContextBySymKey(mechanism, CKA_ENCRYPT, key, param));
+    dec_ctx.reset(PK11_CreateContextBySymKey(mechanism, CKA_DECRYPT, key, param));
```

Comment:
Not sure if CryptoAESKeyHandler is always used for both encryption and decryption. If not, initializing one instead of both contexts might be beneficial.

Reply:
no, it's not always for both encryption and decryption. if it's used by the RotatingKeyRing, it will be only used for deciphering.  but it will complicate the code if we want to differentiate a decode-only CryptoHandler from the bi-direction CryptoHandler: please note `CryptoKey` and `CryptoKeyHandler` support both encryption and decryption. we need to specialize them to enforce this difference. probably we can leave it for another PR.

Reply:
Nevermind then, no need to complicate code futher in that case.